### DISCUSSION TITLE
How to determine the type of an InputEvent?

### DISCUSSION THREAD

**Discussion by ullanar** [quaadgras/graphics.gd#38] [Q&A] [answered]
Hi!

Here is simple code snippet from examples

```go
func (p *Player) Process(delta gd.Float) {
	Input := gd.Input(p.Temporary)
	//...
}
```
`gd.Input` returns `classdb.Input`. I want to create function which handles all inputs and returns action name which should be processed into my state machine. 

So i want to make function like

```go
func (p *Player) HandleInput(event gd.InputEvent) {
       // trying to handle mouse modes
	if event.AsInputEvent().IsActionPressed(p.Temporary.StringName("ui_cancel"), true, true) {
		mouseMode := gd.Input(p.Temporary).GetMouseMode()
		if mouseMode == gd.InputMouseMode(0) {
			gd.Input(p.Temporary).SetMouseMode(gd.InputMouseMode(1))
		} else {
			gd.Input(p.Temporary).SetMouseMode(gd.InputMouseMode(0))
		}
	}

    // rest of strange code to collect movements
}
```

How can I get that `gd.InputEvent` type from `classdb.Input`? And also specific `InputEventMouseMotion` to handle rotation?

**Comment by Splizard** [comment, marked as answer]
You can switch on the underlying event type and access the derived godot class like this:
(I also recommend capturing `Input` and `tmp` variables and using the named enum values to make the code more concise and readable)
```go
func (p *Player) HandleInput(event gd.InputEvent) {
	var (
		tmp   = p.Temporary
		Input = gd.Input(tmp)
	)
	if event.IsActionPressed(tmp.StringName("ui_cancel"), true, true) {
		mouseMode := Input.GetMouseMode()
		if mouseMode == gd.InputMouseModeVisible {
			Input.SetMouseMode(gd.InputMouseModeHidden)
		} else {
			Input.SetMouseMode(gd.InputMouseModeVisible)
		}
	}
	// Switch on the underlying event type.
	switch event := tmp.Variant(event).Interface(tmp).(type) {
	case gd.InputEventMouseMotion:

	case gd.InputEventMouseButton:

	}
}
```
I understand it's not obvious how to do the type switch, so I appreciate you reporting your experience here and I will be considering ideas on how to make this more intuitive.

**Reply by ullanar** [reply]
Thank you! Also in question I wanted to ask about `gd.InputEvent` type. In `HandleInput` in this case we need to accept event param. What is the way to get it?

**Reply by Splizard** [reply]
Implement the `UnhandledInput(gd.InputMethod)` method on a Node.

**Comment by TTSKarlsson** [comment]
Is there an update to this as `.Temporary` doesn't seem to exist anymore?

**Reply by Splizard** [reply]
You can use `Object.Is` and `Object.As` to check and convert to a specific object type.
https://pkg.go.dev/graphics.gd/variant/Object

**Comment by TTSKarlsson** [comment]
Thank you!
